### PR TITLE
EN-4452 : Staking SC - unit tests

### PR DESCRIPTION
**Pull request by miiu96**
Added unit tests for staking sc and fix some small problems

### CODE REVIEW

File: vm/systemSmartContracts/staking_test.go [hunk 1]
```diff
@@ -1 +1,448 @@
...
+	assert.NotNil(t, stakingSmartContract)
+	assert.Nil(t, err)
+}
+
+func TestStakingSC_ExecuteInit(t *testing.T) {
+	t.Parallel()
+
+	stakeValue := big.NewInt(100)
+	eei, _ := NewVMContext(&mock.BlockChainHookStub{}, &mock.CryptoHookStub{})
+	eei.SetSCAddress([]byte("addr"))
```
[the remark this answers is not in the document]
Comment (camilbancioiu) [reply vm/systemSmartContracts/staking_test.go]: `eei.SetSCAddress()` is a step that needs to be performed on the EEI before calling any function of a SC. Because the EEI will only allow a SmartContract to access its own state exclusively, the EEI must be informed of the address of the SmartContract that is to be called. When the SmartContract later calls `GetStorage()`, the EEI will retrieve the storage under that specific SC address. This is why calling `eei.SetSCAddress()` happens *before* the SC is called. Normally, preparing the EEI is a step performed by the VM automatically, when a SC call is requested.

File: vm/systemSmartContracts/staking.go [hunk 2]
```diff
@@ -162,7 +170,7 @@ func (r *stakingSC) finalizeUnStake(args *vmcommon.ContractCallInput) vmcommon.R
 	var registrationData stakingData
 	for _, arg := range args.Arguments {
 		data := r.eei.GetStorage(arg.Bytes())
-		err := json.Unmarshal(data, registrationData)
+		err := json.Unmarshal(data, &registrationData)
```
[the remark this answers is not in the document]
Comment (miiu96) [reply vm/systemSmartContracts/staking.go]: If data is nil you will get an UserError because Unmarshal cannot be done.

File: vm/systemSmartContracts/staking_test.go [hunk 3]
```diff
@@ -1 +1,448 @@
...
+	stakingSmartContract, _ := NewStakingSmartContract(stakeValue, eei)
+	arguments := CreateVmContractCallInput()
+	arguments.Function = "_init"
+
+	retCode := stakingSmartContract.Execute(arguments)
+	assert.Equal(t, vmcommon.Ok, retCode)
+
+	data := stakingSmartContract.eei.GetStorage(arguments.CallerAddr)
```
[the remark this answers is not in the document]
Comment (miiu96) [reply vm/systemSmartContracts/staking_test.go]: Done. Now i check if CallerAddr is put  in storage.

File: vm/systemSmartContracts/staking_test.go [hunk 4]
```diff
@@ -1 +1,448 @@
...
+	arguments := CreateVmContractCallInput()
+	arguments.CallerAddr = []byte("data")
+	arguments.Function = "finalizeUnStake"
+	arguments.Arguments = []*big.Int{big.NewInt(100)}
+
+	retCode := stakingSmartContract.Execute(arguments)
+	assert.Equal(t, vmcommon.UserError, retCode)
+}
+
+func TestStakingSC_ExecuteFinalizeUnStake(t *testing.T) {
```
Comment (camilbancioiu) [inline vm/systemSmartContracts/staking_test.go]: Suggestion: create a test that performs both the `unstake` and `finalizeUnstake` operations in succession. This test simulates a prior `unstake`, then performs `finalizeUnstake`.

File: vm/systemSmartContracts/staking_test.go [hunk 5]
```diff
@@ -1 +1,448 @@
...
+	t.Parallel()
+
+	registrationData := stakingData{
+		StartNonce:    0,
+		Staked:        true,
+		UnStakedNonce: 10,
+		BlsPubKey:     nil,
+		StakeValue:    big.NewInt(100),
+	}
+	argument := big.NewInt(100)
```
Comment (camilbancioiu) [inline vm/systemSmartContracts/staking_test.go]: Rename to `blsPubKey`

File: vm/systemSmartContracts/staking_test.go [hunk 6]
```diff
@@ -1 +1,448 @@
...
+	t.Parallel()
+
+	stakeValue := big.NewInt(0)
+	eei := &mock.SystemEIStub{}
+	eei.GetStorageCalled = func(key []byte) []byte {
+		switch {
+		case bytes.Equal(key, []byte(ownerKey)):
+			return []byte("data")
+		default:
+			registrationDataMarshalized, _ := json.Marshal(&stakingData{StakeValue: big.NewInt(100)})
```
Comment (camilbancioiu) [inline vm/systemSmartContracts/staking_test.go]: Shouldn't `stakingData` also contain `Staked: true`?

File: vm/systemSmartContracts/staking_test.go [hunk 7]
```diff
@@ -1 +1,448 @@
...
+	arguments := CreateVmContractCallInput()
+	arguments.Function = "slash"
+	arguments.CallerAddr = []byte("data")
+	arguments.Arguments = []*big.Int{big.NewInt(100), big.NewInt(100)}
+
+	retCode := stakingSmartContract.Execute(arguments)
+	assert.Equal(t, vmcommon.UserError, retCode)
+}
+
+func TestStakingSC_ExecuteSlash(t *testing.T) {
```
Comment (camilbancioiu) [inline vm/systemSmartContracts/staking_test.go]: Suggestion: write a test which verifies that only an address with a stake can be slashed, and an address that has no stake cannot be slashed.

File: vm/systemSmartContracts/staking.go [hunk 8]
```diff
@@ -86,7 +89,7 @@ func (r *stakingSC) stake(args *vmcommon.ContractCallInput) vmcommon.ReturnCode
 	data := r.eei.GetStorage(args.CallerAddr)
```
[the remark this answers is not in the document]
Comment (miiu96) [reply vm/systemSmartContracts/staking.go]: Stake value will be save in storage.

File: vm/systemSmartContracts/staking_test.go [hunk 9]
```diff
@@ -1 +1,448 @@
...
+
+	var registrationData stakingData
+	data := stakingSmartContract.eei.GetStorage(arguments.CallerAddr)
+	err := json.Unmarshal(data, &registrationData)
+	assert.Nil(t, err)
+	assert.Equal(t, expectedRegistrationData, registrationData)
+
+	receiverBalance := stakingSmartContract.eei.GetBalance(arguments.RecipientAddr)
+	senderBalance := stakingSmartContract.eei.GetBalance(arguments.CallerAddr)
+	assert.Equal(t, big.NewInt(-100), senderBalance)
```
[the remark this answers is not in the document]
Comment (miiu96) [reply vm/systemSmartContracts/staking_test.go]: why -100 is not ok because when you transfer balance if account not exist will be created an account with 0 balance and from 0 is subtract 100 and sender balance will be -100 ?

File: vm/systemSmartContracts/staking_test.go (hunk 5 above)
Comment (miiu96) [reply vm/systemSmartContracts/staking_test.go]: done.

File: vm/systemSmartContracts/staking_test.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (sasurobert) [reply vm/systemSmartContracts/staking_test.go]: -100 is ok. As these are local changes. Not actual account balance values.

File: vm/systemSmartContracts/staking_test.go [hunk 10]
```diff
@@ -1 +1,448 @@
...
+	eei := &mock.SystemEIStub{}
+	eei.GetStorageCalled = func(key []byte) []byte {
+		return []byte("data")
+	}
+	stakingSmartContract, _ := NewStakingSmartContract(stakeValue, eei)
+	arguments := CreateVmContractCallInput()
+	arguments.Function = "stake"
+
+	retCode := stakingSmartContract.Execute(arguments)
+	assert.Equal(t, vmcommon.UserError, retCode)
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply vm/systemSmartContracts/staking_test.go]: Other type of return might be used, but we do not have a lot :D

File: vm/systemSmartContracts/staking_test.go [hunk 11]
```diff
@@ -1 +1,448 @@
...
+}
+
+func TestStakingSC_ExecuteStakeNotEnoughArgsShouldErr(t *testing.T) {
+	t.Parallel()
+
+	stakeValue := big.NewInt(0)
+	eei := &mock.SystemEIStub{}
+	eei.GetStorageCalled = func(key []byte) []byte {
+		registrationDataMarshalized, _ := json.Marshal(&stakingData{})
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply vm/systemSmartContracts/staking_test.go]: yes

File: vm/systemSmartContracts/staking_test.go [hunk 12]
```diff
@@ -1 +1,448 @@
...
+	}
+
+	stakeValue := big.NewInt(100)
+	eei, _ := NewVMContext(&mock.BlockChainHookStub{}, &mock.CryptoHookStub{})
+	eei.SetSCAddress([]byte("addr"))
+
+	stakingSmartContract, _ := NewStakingSmartContract(stakeValue, eei)
+	arguments := CreateVmContractCallInput()
+	arguments.Function = "stake"
+	arguments.Arguments = []*big.Int{big.NewInt(100), big.NewInt(200)}
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply vm/systemSmartContracts/staking_test.go]: yes

File: vm/systemSmartContracts/staking_test.go (hunk 6 above)
Comment (miiu96) [reply vm/systemSmartContracts/staking_test.go]: yes

File: vm/systemSmartContracts/staking_test.go [hunk 13]
```diff
@@ -1 +1,448 @@
...
+	t.Parallel()
+
+	stakeValue := big.NewInt(100)
+	eei := &mock.SystemEIStub{}
+	stakingSmartContract, _ := NewStakingSmartContract(stakeValue, eei)
+	arguments := CreateVmContractCallInput()
+	arguments.Function = "stake"
+
+	retCode := stakingSmartContract.Execute(arguments)
+	assert.Equal(t, vmcommon.UserError, retCode)
```
[the remark this answers is not in the document]
Comment (miiu96) [reply vm/systemSmartContracts/staking_test.go]: Done.

File: vm/systemSmartContracts/staking_test.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (raduchis) [reply vm/systemSmartContracts/staking_test.go]: ok

File: vm/systemSmartContracts/staking_test.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (raduchis) [reply vm/systemSmartContracts/staking_test.go]: there should never exist negative balances. If I stake from an account that does not exist, should the stakeing work?

[the remark this answers is not in the document]
Comment (camilbancioiu) [reply vm/systemSmartContracts/staking_test.go]: But `eei.GetBalance()` returns the real balance of an account. According to the local Node, indeed, but it's still a real balance in ERD, right?

File: vm/systemSmartContracts/staking_test.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (camilbancioiu) [reply vm/systemSmartContracts/staking_test.go]: @miiu96 can you declare variables for the big ints given to `arguments.Arguments`? it will make it clearer what is what.

File: vm/systemSmartContracts/staking_test.go [hunk 14]
```diff
@@ -1 +1,543 @@
...
+	}
+
+	stakingSmartContract, _ := NewStakingSmartContract(stakeValue, eei)
+	arguments := CreateVmContractCallInput()
+	arguments.Function = "slash"
+	arguments.CallerAddr = []byte("data")
+	arguments.Arguments = []*big.Int{big.NewInt(100), big.NewInt(100)}
+
+	retCode := stakingSmartContract.Execute(arguments)
+	assert.Equal(t, vmcommon.Ok, retCode)
```
Comment (camilbancioiu) [inline vm/systemSmartContracts/staking_test.go]: You could also assert that the user's stake is now lower after slashing, equal to an expected value.

File: vm/systemSmartContracts/staking_test.go (hunk 6 above)
Comment (camilbancioiu) [reply vm/systemSmartContracts/staking_test.go]: Please add it.

File: vm/systemSmartContracts/staking.go [hunk 15]
```diff
@@ -198,19 +211,31 @@ func (r *stakingSC) slash(args *vmcommon.ContractCallInput) vmcommon.ReturnCode
...
 	}
 
-	if len(data) == 0 {
-		log.Error("slash error: validator was not registered")
-		return vmcommon.UserError
+	if !registrationData.Staked {
+		return vmcommon.Ok
 	}
 
 	operation := big.NewInt(0).Set(registrationData.StakeValue)
```
Comment (raduchis) [inline vm/systemSmartContracts/staking.go]: operation is not a good name for this. call it stakeValue

File: vm/systemSmartContracts/staking.go [hunk 16]
```diff
@@ -198,19 +211,31 @@ func (r *stakingSC) slash(args *vmcommon.ContractCallInput) vmcommon.ReturnCode
...
 	registrationData.StakeValue = registrationData.StakeValue.Sub(operation, args.Arguments[1])
```
Comment (raduchis) [inline vm/systemSmartContracts/staking.go]: StakeValue could be negative. What to do when slashing is bigger than the current StakeValue that remains. Example min stake is 100, slash 2 times with 70.

File: vm/systemSmartContracts/staking_test.go (hunk 6 above)
Comment (miiu96) [reply vm/systemSmartContracts/staking_test.go]: Done.

File: vm/systemSmartContracts/staking.go [hunk 17]
```diff
@@ -198,19 +211,31 @@ func (r *stakingSC) slash(args *vmcommon.ContractCallInput) vmcommon.ReturnCode
...
+	err := json.Unmarshal(data, &registrationData)
 	if err != nil {
 		log.Error("unmarshal error on slash function" + err.Error())
 		return vmcommon.UserError
 	}
 
-	if len(data) == 0 {
-		log.Error("slash error: validator was not registered")
-		return vmcommon.UserError
+	if !registrationData.Staked {
```
[the remark this answers is not in the document]
Comment (miiu96) [reply vm/systemSmartContracts/staking.go]: You right.  It is not ok to return vcommon.OK. In this case an UserError needs to be returned because you cannot slash a user that not staked money or already unstake.

File: vm/systemSmartContracts/staking.go [hunk 18]
```diff
@@ -198,19 +211,31 @@ func (r *stakingSC) slash(args *vmcommon.ContractCallInput) vmcommon.ReturnCode
 
 	var registrationData stakingData
 	data := r.eei.GetStorage(args.Arguments[0].Bytes())
```
[the remark this answers is not in the document]
Comment (miiu96) [reply vm/systemSmartContracts/staking.go]: Done.

File: vm/systemSmartContracts/staking.go (hunk 16 above)
[the remark this answers is not in the document]
Comment (miiu96) [reply vm/systemSmartContracts/staking.go]: Done.

File: vm/systemSmartContracts/staking.go (hunk 15 above)
Comment (miiu96) [reply vm/systemSmartContracts/staking.go]: Done.

File: vm/systemSmartContracts/staking.go (hunk 16 above)
Comment (miiu96) [reply vm/systemSmartContracts/staking.go]: It is ok value to be negative.

File: vm/systemSmartContracts/staking_test.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (miiu96) [reply vm/systemSmartContracts/staking_test.go]: Done.

File: vm/systemSmartContracts/staking.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (miiu96) [reply vm/systemSmartContracts/staking.go]: Done.

File: vm/systemSmartContracts/staking.go [hunk 19]
```diff
@@ -197,20 +213,35 @@ func (r *stakingSC) slash(args *vmcommon.ContractCallInput) vmcommon.ReturnCode
...
+		log.Error("cannot slash already unstaked or user not staked")
 		return vmcommon.UserError
 	}
 
-	operation := big.NewInt(0).Set(registrationData.StakeValue)
-	registrationData.StakeValue = registrationData.StakeValue.Sub(operation, args.Arguments[1])
+	stakedValue := big.NewInt(0).Set(registrationData.StakeValue)
+	slashValue := args.Arguments[1]
+	registrationData.StakeValue = registrationData.StakeValue.Sub(stakedValue, slashValue)
+	registrationData.Staked = false
```
Comment (raduchis) [inline vm/systemSmartContracts/staking.go]: Should the Staked be set to false after slashing? This means that if we slash just a little of the stake the validator will not be able to validate any longer. 
Please add this test in the TestStakingSC_ExecuteSlashStaked, that the Staked is false for the slashed account.

Also add a test to do two times slashing of 70 with a minimum stake of 100. See what happens. Should the second one return ok or err?

File: vm/systemSmartContracts/staking_test.go [hunk 20]
```diff
@@ -1 +1,448 @@
...
+
+	stakeValue := big.NewInt(100)
+	eei := &mock.SystemEIStub{}
+	stakingSmartContract, err := NewStakingSmartContract(stakeValue, eei)
+
+	assert.NotNil(t, stakingSmartContract)
+	assert.Nil(t, err)
+}
+
+func TestStakingSC_ExecuteInit(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (raduchis) [reply vm/systemSmartContracts/staking_test.go]: This is not yet fixed.

File: vm/systemSmartContracts/staking_test.go [hunk 21]
```diff
@@ -1 +1,448 @@
...
+	arguments.Function = "_init"
+
+	retCode := stakingSmartContract.Execute(arguments)
+	assert.Equal(t, vmcommon.Ok, retCode)
+
+	data := stakingSmartContract.eei.GetStorage(arguments.CallerAddr)
+	assert.Equal(t, 0, len(data))
+}
+
+func TestStakingSC_ExecuteStakeWrongStakeValueShouldErr(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (raduchis) [reply vm/systemSmartContracts/staking_test.go]: This is not yet done.

File: vm/systemSmartContracts/staking.go [hunk 22]
```diff
@@ -32,6 +33,9 @@ func NewStakingSmartContract(stakeValue *big.Int, eei vm.SystemEI) (*stakingSC,
 	if stakeValue == nil {
 		return nil, vm.ErrNilInitialStakeValue
 	}
+	if stakeValue.Cmp(big.NewInt(0)) == -1 {
```
[the remark this answers is not in the document]
Comment (miiu96) [reply vm/systemSmartContracts/staking.go]: Condition was modified. Stake value must be greater than zero.

File: vm/systemSmartContracts/staking_test.go (hunk 20 above)
[the remark this answers is not in the document]
Comment (miiu96) [reply vm/systemSmartContracts/staking_test.go]: Done. On init function a check was added that verify if in storage already exist owner address.

[the remark this answers is not in the document]
Comment (miiu96) [reply vm/systemSmartContracts/staking_test.go]: And also o test was created.

File: vm/systemSmartContracts/staking_test.go (hunk 21 above)
[the remark this answers is not in the document]
Comment (miiu96) [reply vm/systemSmartContracts/staking_test.go]: This check is done on a upper level.

File: vm/systemSmartContracts/staking.go (hunk 19 above)
Comment (miiu96) [reply vm/systemSmartContracts/staking.go]: After a discussion with Robert, Staked should not change it value on slash. So i remove the  code where Staked is set to false on slash.

Comment (miiu96) [reply vm/systemSmartContracts/staking.go]: The second  slash for the moment should not return an error. For the moment slash function know just to subtract the slash value from the staked value and save in storage modifications.

File: vm/systemSmartContracts/staking_test.go (hunk 20 above)
[the remark this answers is not in the document]
Comment (raduchis) [reply vm/systemSmartContracts/staking_test.go]: ok

File: vm/systemSmartContracts/staking_test.go (hunk 21 above)
[the remark this answers is not in the document]
Comment (raduchis) [reply vm/systemSmartContracts/staking_test.go]: ok

File: vm/systemSmartContracts/staking.go (hunk 16 above)
[the remark this answers is not in the document]
Comment (raduchis) [reply vm/systemSmartContracts/staking.go]: ok

File: vm/systemSmartContracts/staking.go (hunk 17 above)
[the remark this answers is not in the document]
Comment (raduchis) [reply vm/systemSmartContracts/staking.go]: ok
